KEP-3619: Add `SupplementalGroupsPolicy` feature fields in Kubernetes API(`Node.Status`) and CRI(`RuntimeStatusResponse`) (#4728)
kubernetes/enhancements: keps/sig-node/3619-supplemental-groups-policy/README.md

To provide users/administrators to know which identities are actually attached to the container process, it proposes to introduce new `User` field in `ContainerStatus`. `User` is an object which consists of `Uid`, `Gid`, `SupplementalGroups` fields for linux containers. This will help users to identify unexpected identities. This field is derived by CRI response (See [user in ContainerStatus](#user-in-containerstatus-1) section).

#### NodeFeatures in NodeStatus which contains SupplementalGroupsPolicy field

Because the actual control(calculation) of supplementary groups to be attached to the first container process will happen inside of CRI implementations (container runtimes), It proposes to add `NodeFeatures` field in `NodeStatus` which contains the `SupplementalGroupsPolicy` feature field in side of it like below so that kubernetes can correctly understand whether underlying CRI implementation implements the feature ot not. The field is assumed drived by CRI response.

```golang
type NodeStatus struct {

> 	// Features describes the set of implemented features implemented by the CRI implementation.

Reviewer: ```suggestion
	// Features describes the set of features implemented by the CRI implementation.
```

	Features *NodeFeatures
}
type NodeFeatures struct {
	// SupplementalGroupsPolicy is set to true if the runtime supports SupplementalGroupsPolicy and ContainerUser.
	SupplementalGroupsPolicy *bool
}
```

> Recently [KEP-3857: Recursive Read-only (RRO) mounts](https://kep.k8s.io/3857) introduced `RuntimeHandlers[].Features`. But this does not fit to use for this KEP because RRO mounts should require to inspect [the OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md) to understand low-level OCI runtime supports RRO or not. However, for this KEP(SupplementalGroupsPolicy), it does not need to inspect [the OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md) because this KEP only affects to [`Process.User.additionalGid`](https://github.com/opencontainers/runtime-spec/blob/main/config.md#user) and this does not depend on [the OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md). So, introducing new `NodeFeatures` in `NodeStatus` does not make any confusion with `RuntimeHandlerFeatures` because we can clearly define how to use them as below:

Reviewer: ```suggestion
Recently [KEP-3857: Recursive Read-only (RRO) mounts](https://kep.k8s.io/3857) introduced `RuntimeHandlers[].Features`. But it is not fit to use for this KEP because RRO mounts requires inspecting [the OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md) to understand whether the low-level OCI runtime supports RRO or not. However, for this KEP(SupplementalGroupsPolicy), it does not need to inspect [the OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md) because this KEP only affects  [`Process.User.additionalGid`](https://github.com/opencontainers/runtime-spec/blob/main/config.md#user) and does not depend on [the OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md). So, introducing new `NodeFeatures` in `NodeStatus` does not conflict with `RuntimeHandlerFeatures` as we can clearly define how to use them as below:
```

> Recently [KEP-3857: Recursive Read-only (RRO) mounts](https://kep.k8s.io/3857) introduced `RuntimeHandlers[].Features`. But this does not fit to use for this KEP because RRO mounts should require to inspect [the OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md) to understand low-level OCI runtime supports RRO or not. However, for this KEP(SupplementalGroupsPolicy), it does not need to inspect [the OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md) because this KEP only affects to [`Process.User.additionalGid`](https://github.com/opencontainers/runtime-spec/blob/main/config.md#user) and this does not depend on [the OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md). So, introducing new `RuntimeFeatures` in `NodeStatus` does not make any confusion with `RuntimeHandlerFeatures` because we can clearly define how to use them as below:

Reviewer 2: I find it very odd that we would call this "RuntimeFeatures" and then explicitly scope it NOT to include features of the runtime.  

What I mean is that the distinction you're making between "runtime" and "runtime handler" is very technical and not helpful or interesting to most people.  A distinction without a meaningful difference?

Reviewer 3: I think my issue with it is supplimental groups have been supported by OCI runtime [since their inception](https://github.com/opencontainers/runtime-spec/blame/3633d035e81cfc7439dd36f37e50720bd991203d/config_linux.go#L24) whereas runtime features were added to check whether an oci runtime has a feature available. It would feel weird to plumb the oci runtimes with supplimental group support to their `runtime features` call because they've basically always supported it.

really the problem is "runtime" was chosen for CRI and "runtime handler" for lower level runtimes, causing this confusion. Could `NodeFeature` work?

Reviewer 2: Do we expect that, over time, these features will reach 100%, or that they will forever be optional?  Is there ever a time when we can remove the code that is conditional on these feature indicators, or do we keep them forever?  Can they be treated like feature gates - after some point in time they don't need to be expressed, they are assumed?

> supplimental groups have been supported by OCI runtime since their inception

Supplemental groups have, but this policy has not, right?  Do we think that EVERY runtime handler will properly handle this?

The issues I have here are 1) naming clarity; 2) do we really need 2 different sets of feature indicators?

Taking 2 first, I forgot that `NodeRuntimeHandlerFeatures` is already in a list-map keyed by handler name, meaning the list of features could be different for different handlers.  So while we *could* take the flags from the runtime and then "augment" that with flags from kubelet, it feels awkward.  

I guess I can buy that we really do have 2 lists.  So then to my point 1 - if we can name this so it is clear, that would be a relief.  As someone who might write a client, I need some way to think about which list I should be checking for which information.

Author: > really the problem is "runtime" was chosen for CRI and "runtime handler" for lower level runtimes, causing this confusion. Could NodeFeature work?

_Similar discussion occurred in [this thread](https://github.com/kubernetes/enhancements/pull/4728#discussion_r1643587997) too._ @AkihiroSuda also suggested similar idea.

The reason why I selected "runtime" is that CRI is abbreviation of Container _`Runtime`_ Interface. CRI actually defines [`RuntimeService`](https://github.com/kubernetes/cri-api/blob/3a66d9d8665454a681e90f6c26b6173c75d5ab53/pkg/apis/runtime/v1/api.proto#L33). Thus, I thought just calling _runtime_ for CRI implementation can be ok.

> Do we expect that, over time, these features will reach 100%,

I'd like to expect so.

> So then to my point 1 - if we can name this so it is clear, that would be a relief.

For naming, how about `NodeStatus.Features`(type=`NodeFeatures`) as @AkihiroSuda suggested [here](https://github.com/kubernetes/enhancements/pull/4728#discussion_r1643587997), @haircommander also suggested in [here](https://github.com/kubernetes/enhancements/pull/4728#discussion_r1644477933)??　 Or, how about more direct naming like `ContainerRuntimeFeatures`, `CRIFeatures`??

> if we can name this so it is clear, that would be a relief. As someone who might write a client, I need some way to think about which list I should be checking for which information.

Previously, I proposed this distinction in this [thread](https://github.com/kubernetes/enhancements/pull/4728#discussion_r1639151566). Is this too technical??

- `NodeFeatures` (this kep): focuses on features that do NOT require to inspect [OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md) (i.e. implemented only in high-level container runtimes)
- `RuntimeHandlers[].Features` (kep-3857): focuses on features that require to inspect [OCI runtime spec's Feature](https://github.com/opencontainers/runtime-spec/blob/main/features.md) (i.e. need implemented in low-level container runtimes)

Author: And, let me clarify my main motivation. My main purpose of this API change is that

- to add E2E test for SupplementalGroupsPolicy in k/k in graceful manner(i.e. it skips the test when the CRI implementation does not implement this feature) (my working [PR](https://github.com/kubernetes/kubernetes/pull/125470))
- to introduce test cases SupplementalGroupsPolicy in critest (my working [PR](https://github.com/kubernetes-sigs/cri-tools/pull/1438))
  - the test cases also needs to be graceful(can skip when CRI does not support the feature) because CRI's CI depends on released version of contaienrd and CRI-O.

Thus, without these API changes proposed in this PR, we might be able to just postpone them (E2E in k/k and update in critest) until the test can be implemented safely in each repos??.

Reviewer 2: The problem with making a distinction between "high-level container runtimes" and "low-level container runtimes" is that there are maybe 20 people in the world who know what that means, and they are all on this PR.  End users and integrators will struggle to comprehend this.  _I_ struggle with it.  Is containerd high-level or low-level?  Is kubelet a "container runtime"?

Given a node with kubelet and containerd, which component will answer the question asked by this field?  (specifically "Is SupplementalGroupsPolicy supported?")

> My main purpose of this API change is ... to add E2E test (i.e. it skips the test when the CRI implementation does not implement this feature

Having e2es that self-disable is a good way to ensure that they are broken .  It's useful, I guess, to let people opt out of them, but we need them to be enabled for our own CI, so that a failure is handled immediately.

Author: > The problem with making a distinction between "high-level container runtimes" and "low-level container runtimes" is that there are maybe 20 people in the world who know what that means, and they are all on this PR. End users and integrators will struggle to comprehend this. I struggle with it. Is containerd high-level or low-level? Is kubelet a "container runtime"?

Yes, totally.

> Given a node with kubelet and containerd, which component will answer the question asked by this field? (specifically "Is SupplementalGroupsPolicy supported?")

As you might know, any single component can't answer this question. This is what makes API design difficult.

- kubelet: it must need to enable feature gate (until GA)
- containerd: it needs to let kubelet know `SupplementalGroupsPolicy` control is implemented or not.
  - _currently we have no such interfaces in CRI, either._ 
  - this is why this PR also proposed to add a interface for this in CRI, too.

> but we need them to be enabled for our own CI, so that a failure is handled immediately.

Do you mean we can add another CI test job for `SupplementalGroupsPolicy` or other CRI implementation dependent alpha features (e.g. RRO)?? I think it's a nice idea, thanks. I would like to research how to do it. _I'm very novis in test-infra._

Reviewer 3: I like plan 1 :) @thockin in your case, containerd would need to be aware. Kubelet is part of kube, and CRI begins the "container runtime" (I like to call them CRI implementations to distinguish between lower level runtimes). the lower level runtimes also need to have support for the field, but they have forever, it's the CRIs that don't yet and we'd want visibility into to have the kubelet trust it was set to report success

Reviewer 2: I'm OK with plan 1

Author: @thockin @haircommander Thank you, guys! I think we agreed to add `NodeStatus.Features` and to add `SupplementalGroupsPolicy` to it.

I renamed `NodeRuntimeFeatures` with `NodeFeatures` in [08fb7f7](https://github.com/kubernetes/enhancements/pull/4728/commits/08fb7f7af32b16570d3b1b54126e8729741c275e).

> - `RuntimeFeatures`(added in this KEP):

Reviewer 4: `NodeStatus.Features` may sound less confusing?

Author: Let me resolve this suggestion because naming discussion resolved in [this thread](https://github.com/kubernetes/enhancements/pull/4728#discussion_r1645280217) as your suggestion.

>   - focses on features that depend only on cri implementation, be independent on runtime handlers(low-level container runtimes), (i.e. it should not require to inspect to any information from oci runtime-spec's features).

Reviewer: ```suggestion
  - focusses on features that depend only on cri implementation, be independent of runtime handlers(low-level container runtimes), (i.e. it should not require to inspect to any information from oci runtime-spec's features).
```

}
```

#### supplemental_groups_policy in RuntimeHandlerFeatures

> To propagate whether runtime supports fine-grained supplemental group control to `NodeRuntimeHandlerFeatures.SupplementalGroupsPolicy`, it proposes to add a corresponding field in `RuntimeHandlerFeatures`. This will also help implementing tests which interacts with CRI so that it can gracefully skip when the container runtime does not support fine-grained supplemental groups control, for example in critest.

Reviewer 3: I don't know if runtime handler features is the correct level, as it's not something the OCI runtime will toggle support for, but the CRI implementation. I wonder if we need a new field in status for this?

Author: >  as it's not something the OCI runtime will toggle support for, but the CRI implementation

Yeah, that makes sense. Then, how about introducing `RuntimeFeatures` in `RuntimeStatus` like this?

```proto
// service RuntimeService {
//     rpc Status(StatusRequest) returns (StatusResponse) {}
// }
// message StatusResponse {
//     RuntimeStatus status = 1;
// }
message RuntimeStatus {
...
    // NEW FIELD
    RuntimeFeatures features = 2;
}
message RuntimeFeatures {
    bool supplemental_groups_policy = 1;  
}
```

> I wonder if we need a new field in status for this?

The main purpose is that I would like to merge below before container runtime releases or backports(I'm not sure this can be approved or not):
- E2E: https://github.com/kubernetes/kubernetes/pull/125470
- critest: https://github.com/kubernetes-sigs/cri-tools/pull/1438

Reviewer 3: works for me though we'll need API review for the node status pieces. WE should make sure in the doc text of the API/CRI to make clear the distinction between the two *Features

Author: I'm starting to wonder if it's really safe to add this at the CRI/NodeRuntime level. Since each container is actually created by OCI runtimes, it’s something to think about.

This KEP only affects the [proc.user.additionalGids](https://github.com/opencontainers/runtime-spec/blob/main/config.md#user) field. This field is pretty important for POSIX platforms, so we can assume most OCI runtimes support it. But what if there are some OCI runtimes that don't?

Reviewer 3: yeah LGTM!

Author: Addressed in 
[a998bd3](https://github.com/kubernetes/enhancements/pull/4728/commits/a998bd3a260c7a60e2d73f42c274ea1942adc0ed)

Author: Let me resolve this suggestion because further discussion went in [this thread](https://github.com/kubernetes/enhancements/pull/4728#discussion_r1643548866).

}
```

#### features in StatusResponse which contains supplemental_groups_policy field

To propagate whether the runtime supports fine-grained supplemental group control to `NodeFeatures.SupplementalGroupsPolicy`, it proposes to add a corresponding field`features` in `StatusResponse`. 

```proto
// service RuntimeService {
// ...
//     rpc Status(StatusRequest) returns (StatusResponse) {}
// }
message StatusResponse {
...
    // features describes the set of features implemented by the CRI implementation.

>     RuntimeFeatures features = ?;

Reviewer 3: nit: should this be NodeFeatures as well?

Author: In CRI, I would like to use "Runtime" instead of "Node" because CRI does not use "Node" wording. WDYT??

Author: @haircommander fixed in [e054a20](https://github.com/kubernetes/enhancements/pull/4728/commits/e054a20cc6f36a14c500c08db6d79c7dc29a9732)

}
```

#### features in StatusResponse which contains supplemental_groups_policy field

> To propagate whether the runtime supports fine-grained supplemental group control to `NodeFeatures.SupplementalGroupsPolicy`, it proposes to add a corresponding field`features` in `StatusResponse`. 

Reviewer: ```suggestion
To propagate whether the runtime supports fine-grained supplemental group control to `NodeFeatures.SupplementalGroupsPolicy`, it proposes to add a corresponding field `features` in `StatusResponse`. 
```

// }
```

#### NodeFeatures in NodeStatus which contains SupplementalGroupsPolicy field

```golang
type NodeStatus struct {
	// Features describes the set of implemented features implemented by the CRI implementation.
	// +featureGate=SupplementalGroupsPolicy
	// +optional
	Features *NodeFeatures

	// The available runtime handlers.
	// +featureGate=RecursiveReadOnlyMounts
	// +optional

>   RuntimeHandlers []RuntimeHandlers

Reviewer 4: nit: indent

Author: Thanks for the catch! fixed in https://github.com/kubernetes/enhancements/pull/4728/commits/668d4a2eb8078a083adb69cda98c7fc5baf100ce
